SwiftGit2/Repository.swift

```diff
@@ -561,89 +582,159 @@ final public class Repository {
+
+		var oldTree: OpaquePointer? = nil
+		if let oid = oldCommitOid {
+			let result = unsafeTreeForCommitId(oid)
+			guard result.error == nil else {
+				return Result.failure(result.error!)
 			}
+
+			oldTree = result.value
 		}
 
-		var newPointer: OpaquePointer? = nil
-		let newResult = self.treePointer(from: newOid)
-		if let value = newResult.value {
-			newPointer = value
-		} else {
-			return newResult
+		var newTree: OpaquePointer? = nil
+		if let oid = newCommitOid {
+			let result = unsafeTreeForCommitId(oid)
+			guard result.error == nil else {
+				return Result.failure(result.error!)
+			}
+
+			newTree = result.value
 		}
 
-		var unsafeDiff: OpaquePointer? = nil
-		let diffResult = git_diff_tree_to_tree(&unsafeDiff,
+		var diff: OpaquePointer? = nil
+		let diffResult = git_diff_tree_to_tree(&diff,
 																					 self.pointer,
-																					 oldPointer,
-																					 newPointer,
+																					 oldTree,
+																					 newTree,
 																					 nil)
-		guard diffResult == GIT_OK.rawValue,
-			let unwrapDiffResult = unsafeDiff else {
+
+		git_object_free(oldTree)
+		git_object_free(newTree)
+
+		guard diffResult == GIT_OK.rawValue else {
 			return .failure(NSError(gitError: diffResult,
 															pointOfFailure: "git_diff_tree_to_tree"))
 		}
 
-		return .success(unwrapDiffResult)
+		return Result<OpaquePointer, NSError>.success(diff!)
 	}
 
-	private func treePointer(from commitOid: OID) -> Result<OpaquePointer, NSError> {
-		var commit: OpaquePointer? = nil
-		var _oid = commitOid.oid
-		let commitResult = git_object_lookup(&commit, self.pointer, &_oid, GIT_OBJ_COMMIT)
-		guard commitResult == GIT_OK.rawValue else {
-			return .failure(NSError(gitError: commitResult,
-															pointOfFailure: "git_object_lookup"))
+	/// Memory safe
+	private func diff(from oldCommitOid: OID?, to newCommitOid: OID?) -> Result<Diff, NSError> {
+		assert(oldCommitOid != nil || newCommitOid != nil, "It is an error to pass nil for both the oldOid and newOid")
+
+		var oldTree: Tree? = nil
+		if oldCommitOid != nil {
+			let result = safeTreeForCommitId(oldCommitOid!)
+			guard result.error == nil else {
+				return Result<Diff, NSError>.failure(result.error!)
+			}
+			oldTree = result.value
 		}
 
-		var tree: OpaquePointer? = nil
-		let treeResult = git_commit_tree(&tree, commit)
-		guard treeResult == GIT_OK.rawValue else {
-			return .failure(NSError(gitError: treeResult,
-															pointOfFailure: "git_object_lookup"))
+		var newTree: Tree? = nil
+		if newCommitOid != nil {
+			let result = self.safeTreeForCommitId(newCommitOid!)
+			guard result.error == nil else {
+				return Result<Diff, NSError>.failure(result.error!)
+			}
+			newTree = result.value!
 		}
 
-		let result = Result<OpaquePointer, NSError>.success(tree!)
-		git_tree_free(tree)
-		git_commit_free(commit)
-		return result
+		if oldTree != nil && newTree != nil {
```

Comment:
I'm unsure of this triple branch of logic. I was hoping to find a "swifty" way of identifying which tree was (or whether both trees were) supplied, without needing three `if else` branches but couldn't get it. I'm particularly interested if something pops out, I tried several different things and none were as satisfying to me as just doing the branch.